src/analyzer/vulnerabilities/arbitrary_from_in_transferfrom.rs

```diff
@@ -0,0 +1,124 @@
+
+use crate::analyzer::ast::{self, Target};
+
+pub fn _arbitrary_from_in_transferfrom_vulnerability(source_unit: SourceUnit) -> HashSet<Loc> {
+    //Create a new hashset that stores the location of each vulnerability target identified
+    let mut vulnerability_locations: HashSet<Loc> = HashSet::new();
+
+    //Extract the target nodes from the source_unit
+    let target_nodes =
+        ast::extract_target_from_node(Target::FunctionDefinition, source_unit.into());
+
+    //For each target node that was extracted, check for the vulnerability patterns
+    for _node in target_nodes {
+        let contract_part = _node.contract_part().unwrap();
+
+        if let pt::ContractPart::FunctionDefinition(box_fn_definition) = contract_part {
+            // if function has no params or no body, it is probably not affected by this vulnerability.
+            if box_fn_definition.params.is_empty() || box_fn_definition.body.is_none() {
+                continue;
+            }
+
+            // We extract the function parameters to later verify if they are used as the first parameter in the `transferFrom/SafeTransferFrom`.
+            let fn_params = extract_fn_params_as_string(&box_fn_definition);
+
+            if let pt::Statement::Block {
+                loc: _,
+                unchecked: _,
+                statements,
+            } = box_fn_definition.body.unwrap()
+            {
+                // We loop through each body expression to determine if 'transferFrom/SafeTransferFrom' is used.
+                for _statement in statements {
+                    if let pt::Statement::Expression(_, box_expression) = _statement.clone() {
+                        if let pt::Expression::FunctionCall(
+                            loc,
+                            box_fn_expression,
+                            fn_call_params,
+                        ) = box_expression
+                        {
+                            // If a call exists and any of the function parameters is used as the first parameter in the call, we mark it as a vulnerability.
+                            if should_be_added(box_fn_expression, &fn_params, &fn_call_params[0]) {
+                                vulnerability_locations.insert(loc);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    //Return the identified vulnerability locations
+    vulnerability_locations
+}
+
+fn extract_fn_params_as_string(box_fn_definition: &Box<pt::FunctionDefinition>) -> Vec<String> {
+    return box_fn_definition
+        .params
+        .iter()
+        .filter(|v| v.clone().1.is_some())
+        .map(|v| v.clone().1.unwrap().name.unwrap().name)
+        .collect();
+}
+
+fn have_transfer_from_expression(box_fn_expression: Box<pt::Expression>) -> bool {
+    if let pt::Expression::MemberAccess(_, _, member_identifier) = *box_fn_expression {
+        return vec!["safetransferfrom", "transferfrom"]
```

Comment:
Can you use `pub const SAFE_TRANSFER_FROM: &str = "safetransferfrom` and `pub const TRANSFER_FROM: &str = "transferfrom` and use these to check if the function name is either of the two?

Reply:
Done